### PR TITLE
Fix ConvertToMD Q conversion for indirect spectroscopy

### LINKED ISSUE #26965: ConvertToMD bug for indirect geometry
The below script should produce a vertical line, but creates a diagonal line in slice viewer. 

```
Load(Filename=r'mccode.h5', OutputWorkspace='mccode', ErrorBarsSetTo1=True, OutputOnlySummedEventWorkspace=False)
LoadInstrument(Workspace='mccode', Filename='MUSHROOM_Definition.xml', InstrumentName='mushroom', RewriteSpectraMap='True')
ConvertToMD(InputWorkspace='mccode', QDimensions='Q3D', dEAnalysisMode='Indirect', Q3DFrames='Q_lab', OutputWorkspace='moo')
```

The scatter plot on the left is some external code to compute `Q_x` and the RHS is the slice viewer.

Screenshot from 2019-10-07 15-18-17

Setting E_f =E_i in the external code gives a plot similar to Mantid's `ConvertToMD`.

### PR DESCRIPTION
**Pull request by martyngigg**
**Description of work.**

Fixes Q calculation for indirect spectrometers and cleans up some variable naming to make the difference between Direct/Indirect more obvious.

**Report to:** ISIS/Rob Bewley

**To test:**

Obtain the data from the original issue and run the script. The image in the slice viewer should now show a vertical column at `Q=-0.5 Angstroms`.

Fixes #26965 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by AndreiSavici** [comment]
How is this different than just using `Crystallographic` convention? This will cause confusion to people using single crystal spectroscopy on direct geometry instruments. I strongly suggest not merging this PR. Some of the users at SNS have already done single crystal experiments on BASIS with the spectroscopy convention. **Reducing data again should not change their images.** Also, finding the UB from peaks will give different results.

If you really want, add an extra convention for `ISIS indirect spectroscopy`, but then you need to follow where it will break things.

**Comment by martyngigg** [comment]
@AndreiSavici concerns are valid and I shouldn't have made this change to a core algorithm  - a moment of madness. Horace also agrees with Mantid so something more subtle is going in here. Closing this so it won't get merged.

**Comment by AnthonyLim23** [comment]
@DavidFair talked to the scientists and said:

For Quasi-inelastic i.e. close to the elastic line, the error is close to zero. It is therefore not noticeable. 
For deeply inelastic scattering the results produced do not match those from other tools. The equations have been derived from first principles, independently from this PR. The same set of changes were derived as this PR. I think we need to understand why the current version is correct. It does match with Horice, but not in the deeply inelastic case. Therefore, we need to somehow allow for this correction to be included.

File: Framework/MDAlgorithms/src/MDTransfModQ.cpp [hunk 1]
```diff
@@ -166,22 +166,28 @@ the algorithm and false otherwise.
...
+    qx = -m_ex * kFinal;
+    qy = -m_ey * kFinal;
+    qz = m_kFixed - m_ez * kFinal;
   } else {
-    k_tr = sqrt((m_Ei + E_tr) / PhysicalConstants::E_mev_toNeutronWavenumberSq);
+    const double kInitial = sqrt(
+        (m_eFixed + deltaE) / PhysicalConstants::E_mev_toNeutronWavenumberSq);
+    qx = -m_ex * m_kFixed;
+    qy = -m_ey * m_kFixed;
+    qz = kInitial - m_ez * m_kFixed;
```
Comment (AnthonyLim23) [inline Framework/MDAlgorithms/src/MDTransfModQ.cpp]: this seems to be the minus of the original code, unless I have missed something

Comment (martyngigg) [reply Framework/MDAlgorithms/src/MDTransfModQ.cpp]: This was @AndreiSavici 's point and we have an option in mantid already to set the convention to use `kf-ki` instead. Andrei pointed out that we can't change the default and we should already have this covered by enabling/disabling the `Crystallography Convention` option as appropriate in `View->Preferences->Mantid`.

Comment (martyngigg) [reply Framework/MDAlgorithms/src/MDTransfModQ.cpp]: Can we try that?

Comment (AnthonyLim23) [reply Framework/MDAlgorithms/src/MDTransfModQ.cpp]: When I did the maths with @DavidFair yesterday, we got it to be with the minus sign. If the values of qz happen to be close to zero, there is a good chance the error would not be noticed.

Comment (martyngigg) [reply Framework/MDAlgorithms/src/MDTransfModQ.cpp]: It should depend on the convention: Inelastic usually uses `ki-kf` where crystallography uses `kf-ki` or do you just mean the `qz` component and not all of them?

Comment (AnthonyLim23) [reply Framework/MDAlgorithms/src/MDTransfModQ.cpp]: I cant remember which convention we used, my notes are at home. I can check on Monday.

**Comment by abuts** [comment]
Not sure about crystallographic convension, but the issue is that it does not match Horace.
The change
``
        (m_eFixed + deltaE) / PhysicalConstants::E_mev_toNeutronWavenumberSq);
    qx = -m_ex * m_kFixed;
    qy = -m_ey * m_kFixed;
    qz = kInitial - m_ez * m_kFixed;
``
does.

File: Framework/MDAlgorithms/src/MDTransfModQ.cpp [hunk 2]
```diff
@@ -166,22 +166,28 @@ the algorithm and false otherwise.
...
   if (this->m_Emode == Kernel::DeltaEMode::Direct) {
-    k_tr = sqrt((m_Ei - E_tr) / PhysicalConstants::E_mev_toNeutronWavenumberSq);
+    const double kFinal = sqrt((m_eFixed - deltaE) /
+                               PhysicalConstants::E_mev_toNeutronWavenumberSq);
+    qx = -m_ex * kFinal;
+    qy = -m_ey * kFinal;
+    qz = m_kFixed - m_ez * kFinal;
   } else {
-    k_tr = sqrt((m_Ei + E_tr) / PhysicalConstants::E_mev_toNeutronWavenumberSq);
+    const double kInitial = sqrt(
```
Comment (martyngigg) [inline Framework/MDAlgorithms/src/MDTransfModQ.cpp]: @AndreiSavici Can you describe exactly why you think this is wrong? It doesn't just flip the sign from before we think the old formula was just wrong for Indirect. The variable naming meant that `k_i` got set to `kfixed` for indirect. Maybe BASIS operates in a regime where the differences are less prominent, e.g. lower energy transfer?

We've had 3 people derive it separately and come to this answer.

**Review by AndreiSavici — DISMISSED** [review]

**Pull request merged**